device/iot-device-samples/pnp-device-sample/Thermostat/src/main/java/samples/com/microsoft/azure/sdk/iot/Thermostat.java

```diff
@@ -0,0 +1,317 @@
+
+                // Update temperature in 2 steps
+                double step = (targetTemperature - temperature) / 2;
+                for (int i = 1; i <=2; i++) {
+                    temperature = BigDecimal.valueOf(temperature + step).setScale(1, RoundingMode.HALF_UP).doubleValue();
+                    Thread.sleep(5 * 1000);
+                }
+
+                EmbeddedPropertyUpdate completedUpdate = new EmbeddedPropertyUpdate(temperature, StatusCode.COMPLETED.value, property.getVersion(), "Successfully updated target temperature");
+                Property reportedPropertyCompleted = new Property(propertyName, completedUpdate);
+                try {
+                    deviceClient.sendReportedProperties(Collections.singleton(reportedPropertyCompleted));
+                } catch (IOException e) {
+                    throw new RuntimeException("IOException when sending reported property update: ", e);
+                }
+                log.debug("Property: Update - {\"{}\": {}°C} is {}", propertyName, temperature, StatusCode.COMPLETED);
+            } else {
+                log.debug("Property: Received an unrecognized property update from service.");
+            }
+        }
+    }
+
+    @AllArgsConstructor
+    private static class EmbeddedPropertyUpdate {
+        @NonNull
+        @SerializedName("value")
+        public Object value;
+        @NonNull
+        @SerializedName("ac")
+        public Integer ackCode;
+        @NonNull
+        @SerializedName("av")
+        public Integer ackVersion;
+        @SerializedName("ad")
+        public String ackDescription;
+    }
+
+    /**
+     * The callback to handle "getMaxMinReport" command.
+     * This method will returns the max, min and average temperature from the specified time to the current time.
+     */
+    private static class GetMaxMinReportMethodCallback implements DeviceMethodCallback {
+        String commandName = "getMaxMinReport";
+        String formatPattern = "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'";
+
+        @SneakyThrows(ParseException.class)
+        @Override
+        public DeviceMethodData call(String methodName, Object methodData, Object context) {
+            if (methodName.equalsIgnoreCase(commandName)) {
+
+                String jsonRequest = new String((byte[]) methodData, StandardCharsets.UTF_8);
+                JsonObject jsonObject = new Gson().fromJson(jsonRequest, JsonObject.class);
+                String sinceString = jsonObject.get("commandRequest").getAsJsonObject().get("value").getAsString();
+
+                DateFormat format = new SimpleDateFormat(formatPattern);
+                Date since = format.parse(sinceString);
+                log.debug("Command: Received - Generating min, max, avg temperature report since {}.", since);
+
+                double runningTotal = 0;
+                Map<Date, Double> filteredReadings = new HashMap<>();
+                for (Map.Entry<Date, Double> entry : temperatureReadings.entrySet()) {
```

Comment:
I wish we supported Java 8+, we could then use `Streams`. In the absence of that, we are stuck with implementing this logic manually.

Reply:
We can make the sample support Java 8 if you'd like. Our SDK is about to deprecate Java 7 support anyways